chore(ci): setup auto docs publish workflow for release vX.X.X tag

ShenQingchuan opened the pull request:
# Automatic workflow for release docs publish

I wrote these guides in order to give our reviewers more context and information here.

## Why for this PR ?

After publishing a new release of Vite, we must open Netlify UI to trigger docs build & publish manually now.

It's because the contents of `docs` folder are actively updating, and we'll always see the latest content on docs site if we use Netlify's git-based deployment, but actually there's something ahead of current release.

So I'm wondering if there's some ways for setting up an automatically deploy workflow which only targets on `vX.X.X` tag, it's a significant sign.

## Is these changes enough ?

Actually, NO. we should add more configs around.

## First: turn off Netlify builds

Since now this repo would not trigger a new deploy by commit pushing, we can skip this step if it's already done.

1. Go to Netlify UI, site administration page.
2. Open **`Site Setting`** panel
3. Open **`Build & Deploy`** tab on the left side
4. Scroll down to **`Build Settings`**, and click **`Edit Settings`**
5. Select **`Builds --> Stop Builds`**

## Second: setting up secrets

This step may requires @yyx990803, @patak-dev or someone has right to access this repo's setting panel, to create a list of following fields:

1. `VITE_GITHUB_TOKEN`: this is a Github Access Token, may need to be generated by manually creating @yyx990803 

    https://github.com/settings/tokens

2. `NETLIFY_AUTH_TOKEN`: this is a personal access token which is supposed to be generate on Netlify.

    https://app.netlify.com/user/applications#personal-access-tokens

3. `NETLIFY_SITE_ID`: this one could be found at [General | Site settings](https://app.netlify.com/sites/vitejs/settings/general)

    But I could use an open API of Netlify to query for result:

```bash
curl https://api.netlify.com/api/v1/sites/vitejs.dev

// id: 8c66140b-3a31-475e-b0c6-ff6689563820
```

![image](https://user-images.githubusercontent.com/46062972/158433303-a81e9517-7169-4c90-8c41-ccea92d45246.png)

## That's all for configuration.

ShenQingchuan committed:
chore(ci): setup auto docs publish workflow for release vX.X.X tag

patak-dev commented:
@ShenQingchuan in Vitest we have a setup that shows a preview for each commit that affects the docs: https://github.com/vitest-dev/vitest/pull/970. Would the action you used here do the same? @antfu do you have a link to guide us to do the same you did for Vitest here?

antfu commented:
> a link to guide

Actually, it's the default behavior of Netlify - unless you disable it explicitly. I guess it's done by Evan for a reason 👀

patak-dev commented:
Added for discussion in the next meeting, let's check with him then.

ShenQingchuan commented:
> 

No, the setup I made here is for release docs publish.
I know that maybe you @patak-dev also want to keep the preview feature of Netlify and not trigger a production release in daily commits pushing. I'll find it out how to do it !

Could you please let me sit in on the next internal meeting when talking about this issue ( for only once 😊 )

ShenQingchuan committed:
chore(docs): split docs deloy into push and release types

ShenQingchuan commented:
# 2nd Update 

Finally we made it!

What's new? I've removed the config field `production-branch: main` because it would send production build request on Netlify by default ( even with `production-deploy` is actually `false` by default ). So we can only use `production-deploy` with an [expression](https://docs.github.com/en/actions/learn-github-actions/expressions) value instead.

https://github.com/vitejs/vite/blob/419b18e97e9b78704fa77cbe79aaf4fc400db9b6/.github/workflows/deploy-docs-on-netlify.yml#L39

It would generate "Preview" deploy as expected when we modified and pushed anything under `docs/`.

References for your need:
- [Github Actions - Contexts' `github` field](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context)
- [Github Actions - `push` event declaration for `on`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#push)
- [Github Actions - `release` event declaration for `on`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release)
- [Deploy Action Plugin official README](https://github.com/nwtgck/actions-netlify)

ShenQingchuan commented:
> 

And I'm supposed to provide some more information...

the "Preview Build" is indeed a feature of Netlify but it's not by default, it depends on which branch you assign as a "Production branch"

- if you push something to the "Production branch", it triggers a "Production build"
- if you push something to another branch, it triggers a "Preview build"

patak-dev commented:
Sorry for taking so much time @ShenQingchuan, v2.9 is out now, so getting back to this.

Some questions:

1. So IIUC, we aren't trying to get netlify to send as the docs previews. I checked in Vitest, and we are getting them even if we don't touch the docs and that is quite noisy: https://github.com/vitest-dev/vitest/pull/1001.

2. If we delete the [netlify.toml](https://github.com/vitejs/vite/pull/7331/files#diff-ab8f79b68b7adff7a07db953bf453f3c5aa6ade98d2b1b67d8432b36392489ed) file, aren't we going to stop being able to use the Netlify UI? For example the `Retry Deploy`. It may still be ok, we haven't used that button until now.

3. About `VITE_GITHUB_TOKEN`, could we directly use `secrets.GITHUB_TOKEN` that is automatically generated for the action?

4. The current workflow will also deploy the docs for production when we release a plugin. Until now, we are only publishing when releasing Vite core. There have been some times when we released a plugin in the middle of a beta for example. So we should check that the tag of the release corresponds to Vite. We should also avoid releasing for production for a beta release. (unrelated, but I think it would be great to have a public https://beta.vitejs.dev deployment for the betas)

ShenQingchuan commented:
> Sorry for taking so much time @ShenQingchuan, v2.9 is out now, so getting back to this.
> 
> Some questions:
> 
> 1. So IIUC, we aren't trying to get netlify to send as the docs previews. I checked in Vitest, and we are getting them even if we don't touch the docs and that is quite noisy: [fix: vitest cannot read the latest config when it changes vitest-dev/vitest#1001](https://github.com/vitest-dev/vitest/pull/1001).
> 2. If we delete the [netlify.toml](https://github.com/vitejs/vite/pull/7331/files#diff-ab8f79b68b7adff7a07db953bf453f3c5aa6ade98d2b1b67d8432b36392489ed) file, aren't we going to stop being able to use the Netlify UI? For example the `Retry Deploy`. It may still be ok, we haven't used that button until now.
> 3. About `VITE_GITHUB_TOKEN`, could we directly use `secrets.GITHUB_TOKEN` that is automatically generated for the action?
> 4. The current workflow will also deploy the docs for production when we release a plugin. Until now, we are only publishing when releasing Vite core. There have been some times when we released a plugin in the middle of a beta for example. So we should check that the tag of the release corresponds to Vite. We should also avoid releasing for production for a beta release. (unrelated, but I think it would be great to have a public https://beta.vitejs.dev deployment for the betas)

Hi, @patak-dev 

Sorry for take so long to notify that this issue got an update! I would like to answer your questions one by one:

1. Sorry for that I don't really get your point at this one, do you mean that we're not going to set a Netlify preview every commit because that's annoying? IIUC, then my answer is "That's ok", now this PR set up "Preview build" only for **every changes under `docs/` dir. 

2. Yes for "Not able to use the 'Re-trigger build' button in Netlify Admin UI", but it's not because of `netlify.toml` deletion, it's because we need to disable(de-activate) "Netlify Build", this setting will effect whether turn to an auto build for each commit in this repo. We also need to get full control of build to implement our specific "logic" about publishing docs. 

  So I think we may not need Netlify UI anymore as we just use this platform to deploy, and the Github Action plugin [`nwtgck/actions-netlify`](https://github.com/nwtgck/actions-netlify) could meet our needs.

3. I don't know about this feature. Does every action task generate its own Github token? 

4. OK, I got your point. I will think about how to change the setting to match our needs. The main difficulty is how to write a correct "conditional expression" in Github Workflow Configuration for `publish-deploy` field.

ShenQingchuan commented:
For the question 4: I may need to get it more clear ... 

**Do we just need to publish docs on `vX.X.X` tags release ?** @patak-dev

ShenQingchuan committed:
fix: adjust workflow config

only trigger production build for core release

patak-dev commented:
4. Yes, only for `vX.X.X`

3. Yes, and it is available as `secrets.GITHUB_TOKEN` https://docs.github.com/en/actions/security-guides/automatic-token-authentication

1. and 2. I was only checking that I understood the implications and I agree that what you are proposing is good 👍

ShenQingchuan committed:
fix: deploy message

ShenQingchuan committed:
fix: use auto generated GITHUB_TOKEN

ShenQingchuan commented:
@yyx990803 

According to my experiment now, we must enable "auto publishing", otherwise the new deploy generated by Github Action can't be the latest "Production build".

ShenQingchuan committed:
fix: wrong reference

patak-dev commented on .github/workflows/deploy-docs-on-netlify.yml:
```diff
@@ -0,0 +1,47 @@
+        with:
+          publish-dir: "docs/.vitepress/dist"
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          production-deploy: ${{ startsWith(github.ref, 'refs/tags/v') }}
```
This will also match for beta releases, no? That we wanted to avoid 
https://github.com/vitejs/vite/releases/tag/v2.9.0-beta.11

ShenQingchuan replied:
yes，maybe a more specific regular expression ..

ShenQingchuan replied:
I use `contains` which is defined as built-in function 

( docs:  https://docs.github.com/en/actions/learn-github-actions/expressions#contains )

ShenQingchuan committed:
fix: add a more specific check for tag name

patak-dev commented:
@ShenQingchuan I just saw that netlify auto-deploys now has a fixed URL for main: 
- https://main--vitejs.netlify.app/

This is great, as we can use it to check the current HEAD state and as temporal docs during the beta. Maybe we should check that this PR isn't going to disable this feature, or improve on it by directly deploying on every commit to main.vitejs.dev ?

ShenQingchuan commented:
@patak-dev 

Yes, that's another awesome feature of Netlify that I didn't know !

Does it means that: Right now we already have a url to preview our latest docs ?

patak-dev commented:
Yes, check https://main--vitejs.netlify.app/! 🙌🏼 
The UI for netlify changed a bit and now I saw this URL instead of the usual preview one. Or maybe someone touched some general configuration in the Vue netlify account 🤔

ShenQingchuan commented:
@patak-dev 

I found it in Netlify docs, but it seems need to require auto build from a branch. but for `main` branch, it's quite spacial because it's actually our production branch. And it leads to update our release site by every commit.

But I've got your idea about `main.vitejs.dev`, I'd think about how to make it work.

ShenQingchuan commented:
@patak-dev  Maybe we can merge this PR first ?

I haven't found any good idea to "Enable auto publishing preview to `main--vitejs.netlify.app" and "Do not auto publishing main branch's deploy", it seems that Netlify do not support this requirement.

patak-dev commented:
@antfu is using https://docs.netlify.com/domains-https/custom-domains/multiple-domains/#branch-subdomains for VueUse, so there is deploy with the last version for each major: https://v6-7-6.vueuse.org/. In Vite it is usual to apply patches to older versions, so it would be better to have a https://v2.vitejs.dev. Maybe we could still use it and for the last minor in each major we tag it as `v2`

bluwy commented:
Do we still want to go ahead with this? I believe we have disabled auto publishing, but we still manually trigger publishing when we cut a release, and this PR automates it?

ShenQingchuan commented:
I think the problem this PR tried to resolve is staled, so it's closable now.
